TMPDIR cleanup and Docs (#2)
marcnnn/flame_slurm_backend: README.md

# FLAMESlurmBackend

The Flame Slurm Backend allows to use FLAME on SLURM HPC Clusters.

> 

Reviewer: ```suggestion
```

      max: 10,
      max_concurrency: 5,
      idle_shutdown_after: 30_000,
      slurm_job: """
      #!/bin/bash
      #SBATCH -o flame.%j.out
      #SBATCH --nodes=1
      #SBATCH --ntasks-per-node=1
      #SBATCH --time=01:00:00

      export SLURM_FLAME_HOST=$(ip -f inet addr show ib0 | awk '/inet/ {print $2}' | cut -d/ -f1)
      """
    }
  ]
```

> This part of the Job definition defines the Host part of the Flame Child Beam.
> In this case the IP of teh infiniband interface is used so that the beam disribution Protokoll
> is communicating over the infiniband IP interface to allow low latency and high bandwith communication.

Reviewer: ```suggestion
The `slurm_job` defines the Slurm job that will run on each spawned machine.

The `SLURM_FLAME_HOST` environment variable is also explicitly customize to the infiniband interface, which will be used by the Erlang VM Distribution layer for low latency and high bandwidth communication:
```

```bash
export SLURM_FLAME_HOST=$(ip -f inet addr show ib0 | awk '/inet/ {print $2}' | cut -d/ -f1)
```
You need to start the Parent Beam with the same configuration.

Example Livebook start on a CUDA 12.5 Node with CUDNN in $HOME for NX ELXA Backend:

> ```bash

Reviewer: ````suggestion

You will need to start the parent Erlang VM (the one that configures FLAME) with the same configuration. If you are using Livebook, here is a script that starts on a CUDA 12.5 with CUDNN in `$HOME`:

```bash
````

#!/bin/bash
export CUDA=/usr/local/cuda-12.5/
export CUDNN=$HOME/cudnn-linux-x86_64-9.5.0.50_cuda12-archive/
export PATH=$PATH:$CUDA/bin
export CPATH=$CPATH:$CUDNN/include:$CUDA/include
export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:$CUDNN/lib
export MIX_INSTALL_DIR=$WORK/mix-cache
export LB_HF_TOKEN=hf_*****

export SLURM_FLAME_HOST=$(ip -f inet addr show ib0 | awk '/inet/ {print $2}' | cut -d/ -f1)

> export BUMBLEBEE_CACHE_DIR=$WORK/bumblebee/

Reviewer: I think we can remove the HF and Bumblebee stuff, what do you say?

```suggestion

export SLURM_FLAME_HOST=$(ip -f inet addr show ib0 | awk '/inet/ {print $2}' | cut -d/ -f1)

```

If it not scheduled within the timeout the job is canceled to not block ressourcess that are no longer needed.
To be able to run the runner with the correct enviorment cluster specific bash file needs to be created.

# Cleanup

On some clusters TMPDIR is cleaned per Job on others not.

> This is why `TMPDIR` is changed to one based on the Job ID, which is automatically cleared when the Job terminates. More precisely, Slurm is automatically configured to send a SIGTERM signal to FLAME 30 seconds before it terminates the Job, which then deletes the directory.

Reviewer: ```suggestion
We automatically set the TMPDIR to be one based on the Job ID. It is then automatically cleared when the Job terminates.

Slurm is also configured to send a SIGTERM signal to FLAME 30 seconds before it terminates the Job, so it starts performing any cleaning up, and the directory is removed after.
```

Author: Since some Slurm cluster are set up to delete the tmp folder automatically, this part can be changed.
As we discussed:
In addition, Flame cleans up the temporary Files it creates when shut down gracefully now https://github.com/phoenixframework/flame/commit/e64ad84b695a7569a351b7e5717c27db97f2451c

```suggestion
Slurm is configured to send a SIGTERM signal to FLAME 30 seconds before it terminates the Job, so it starts performing any cleaning up of Flame temporary Files.

If your Slurm cluster is not configured to delete the tmp folder you can use OTP supervisors and delete artifacts that you create on termination.

This implementation in Flame is a good reference how to do that:
https://github.com/phoenixframework/flame/commit/e64ad84b695a7569a351b7e5717c27db97f2451c
```

Reviewer: I think we can remove the last two paragraphs, only keep the first one?

If it not scheduled within the timeout the job is canceled to not block ressourcess that are no longer needed.
To be able to run the runner with the correct enviorment cluster specific bash file needs to be created.

# Cleanup

On some clusters TMPDIR is cleaned per Job on others not.

> This is why TMPDIR is changed to one based on the Job ID and deleted after SIGUSR1 is send by SLURM this is Configured by default in Flame SLurm to 30 Seconds before kill.

Reviewer: ```suggestion
This is why `TMPDIR` is changed to one based on the Job ID, which is automatically cleared when the Job terminates. More precisely, Slurm is automatically configured to send a SIGUSR1 signal to FLAME 30 seconds before it terminates the Job, which then deletes the directory.
```

Reviewer: It is not clear to me what happens 30 seconds before kill :)

# Long running Jobs

> If your Job time is Limited Slurm will kill the Job with the Flame runner.
> There are no mechanisims in place to not use the runner if time is about to run out.
> 
> This would be a well appriciated contribution!
> 
> Be aware that you might loose Data because of this.

Reviewer: ```suggestion
If your job time is limited, Slurm will kill the Job while it is running. There are no mechanisms at the moment to not use the runner if time is about to run out. This would be a well appreciated contribution! Be aware that you might lose data because of this.
```
